File: Grafos/Graph.cs
public class Graph
{
    //Numero de vertices que compoe o grafo
    private int numVertices;
    //Lista de adjacência - Grafo representado em uma lista de listas
    //(Cada vertice, tem uma lista de vizinhos adjacentes)
    public List<List<int>> adjList = new List<List<int>>();

    //Representação das arestas em forma de matriz de adjacência
    public bool[,] adjMatrix;


    /// <summary>
    /// Construtor do Grafo
    /// </summary>
    /// <param name="vertices">Numero de vertices que compoe o grafo</param>
    public Graph(int vertices)
    {
        numVertices = vertices;
        adjMatrix = new bool[vertices, vertices];

        for (int i = 0; i < numVertices; i++)
        {
            adjList.Add(new List<int>());
        }
    }

    /// <summary>
    /// Adiciona arestas entre os vertices do grafo
    /// </summary>
    /// <param name="src">Vertice Origem</param>
    /// <param name="dest">Vertice Destino</param>
    public void AddEdge(int src, int dest)
    {
        adjList[src].Add(dest);
        adjList[dest].Add(src);

        adjMatrix[src, dest] = true;
        adjMatrix[dest, src] = true;
    }

    #region BFS
    /// <summary>
    /// Realiza busca em largura na lista de adjacência
    /// </summary>
    /// <param name="startVertice">Vertice para começar a busca</param>
    public void BreadthFirstSearch(int startVertice)
    {
        //Vetor dos vertices - se foram visitados ou nao
        bool[] visited = new bool[numVertices];
        //Fila de vizinhos
        Queue<int> queue = new Queue<int>();

        //Marca o nosso vertice inicial como visitado,
        //e adiciona na fila para visitar os vizinhos dele
        visited[startVertice] = true;
        queue.Enqueue(startVertice);

        while (queue.Count > 0)
        {
            //Pega o primeiro elemento da fila
            int currentVertex = queue.Dequeue();
            Console.Write(currentVertex + " ");

            //Visita dos vizinhos desse elemento e adiciona na fila
            foreach (int adj in adjList[currentVertex])
            {
                if (!visited[adj])
                {
                    visited[adj] = true;
                    queue.Enqueue(adj);
                }
            }
        }
    }
    #endregion

    #region DFS
    /// <summary>
    /// Realiza busca em profundidade na lista de adjacência utilizando uma pilha
    /// </summary>
    /// <param name="startVertice">Vertice para começar a busca</param>
    public void DepthFirstSearch(int startVertice)
    {
        // Vetor dos vértices - se foram visitados ou não
        bool[] visited = new bool[numVertices];

        // Pilha para simular o comportamento recursivo
        Stack<int> stack = new Stack<int>();

        // Adiciona o vértice inicial na pilha
        stack.Push(startVertice);

        while (stack.Count > 0)
        {
            // Pega o topo da pilha (último vértice inserido)
            int currentVertex = stack.Pop();

            // Se o vértice já foi visitado, continua para o próximo
            if (visited[currentVertex])
                continue;

            // Marca o vértice como visitado e imprime
            visited[currentVertex] = true;
            Console.Write(currentVertex + " ");

            // Adiciona os vizinhos não visitados à pilha
            // Os vizinhos são inseridos na pilha na ordem inversa para que o primeiro da lista seja visitado primeiro
            foreach (int adj in adjList[currentVertex])
            {
                if (!visited[adj])
                {
                    stack.Push(adj);
                }
            }
        }
    }


    /// <summary>
    /// Realiza busca em profundidade na lista de adjacência utilizando recursividade
    /// </summary>
    /// <param name="startVertice">Vertice para começar a busca</param>
    public void DepthFirstSearchRecursive(int startVertice)
    {
        // Vetor dos vértices - se foram visitados ou não
        bool[] visited = new bool[numVertices];

        // Chama o método auxiliar recursivo
        DFSRecursive(startVertice, visited);
    }

    /// <summary>
    /// Método auxiliar recursivo para realizar a busca em profundidade
    /// </summary>
    /// <param name="vertex">Vértice atual</param>
    /// <param name="visited">Vetor de vértices visitados</param>
    private void DFSRecursive(int vertex, bool[] visited)
    {
        // Marca o vértice como visitado
        visited[vertex] = true;

        // Exibe o vértice visitado
        Console.Write(vertex + " ");

        // Visita todos os vizinhos não visitados
        foreach (int adj in adjList[vertex])
        {
            if (!visited[adj])
            {
                DFSRecursive(adj, visited);
            }
        }
    }
    #endregion
}

File: Grafos/C/src/graph.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include "graph.h"

// Função para criar um grafo com o número de vértices especificado
Graph* createGraph(int vertices) {
    Graph* graph = (Graph*)malloc(sizeof(Graph));
    graph->numVertices = vertices;

    for (int i = 0; i < vertices; i++) {
        graph->adjList[i] = (int*)malloc(vertices * sizeof(int));  // Aloca memória para a lista de adjacência
        for (int j = 0; j < vertices; j++) {
            graph->adjMatrix[i][j] = false;  // Inicializa a matriz de adjacência
        }
    }

    return graph;
}

// Função para adicionar uma aresta entre dois vértices
void addEdge(Graph* graph, int src, int dest) {
    graph->adjList[src][dest] = 1;
    graph->adjList[dest][src] = 1;

    graph->adjMatrix[src][dest] = true;
    graph->adjMatrix[dest][src] = true;
}

// Função BFS (Busca em Largura)
void bfs(Graph* graph, int startVertex, bool* visited) {
    Queue* queue = createQueue();
    visited[startVertex] = true;
    enqueue(queue, startVertex);

    while (!isQueueEmpty(queue)) {
        int currentVertex = dequeue(queue);
        printf("%d ", currentVertex);

        // Visita todos os vértices adjacentes
        for (int i = 0; i < graph->numVertices; i++) {
            if (graph->adjMatrix[currentVertex][i] && !visited[i]) {
                visited[i] = true;
                enqueue(queue, i);
            }
        }
    }

    freeQueue(queue);
}

// Função DFS (Busca em Profundidade)
void dfs(Graph* graph, int startVertex, bool* visited) {
    Stack* stack = createStack();
    push(stack, startVertex);

    while (!isStackEmpty(stack)) {
        int currentVertex = pop(stack);
        if (visited[currentVertex]) continue;

        visited[currentVertex] = true;
        printf("%d ", currentVertex);

        // Adiciona os vértices adjacentes na pilha
        for (int i = 0; i < graph->numVertices; i++) {
            if (graph->adjMatrix[currentVertex][i] && !visited[i]) {
                push(stack, i);
            }
        }
    }

    freeStack(stack);
}

// Função auxiliar recursiva para DFS
void dfsRecursive(Graph* graph, int vertex, bool* visited) {
    visited[vertex] = true;
    printf("%d ", vertex);

    for (int i = 0; i < graph->numVertices; i++) {
        if (graph->adjMatrix[vertex][i] && !visited[i]) {
            dfsRecursive(graph, i, visited);
        }
    }
}

// Função DFS Recursiva
void depthFirstSearchRecursive(Graph* graph, int startVertex) {
    bool visited[MAX_VERTICES] = {false};
    dfsRecursive(graph, startVertex, visited);
}

// Função para verificar se o grafo é conectado
bool isConnected(Graph* graph) {
    bool visited[MAX_VERTICES] = {false};

    bfs(graph, 0, visited);

    for (int i = 0; i < graph->numVertices; i++) {
        if (!visited[i]) {
            return false;
        }
    }
    return true;
}
